kalatori-ah/src/rpc.rs

```diff
@@ -0,0 +1,1352 @@
+                format!("expected `u128` as the type of a balance, got {encoded_balance}")
+            })
+        } else {
+            Ok(0)
+        }
+    }
+
+    async fn batch_transfer(
+        &self,
+        nonce: Nonce,
+        block_hash_count: BlockNumber,
+        signer: &PairSigner<RuntimeConfig, Pair>,
+        transfers: Vec<Value>,
+    ) -> Result<SubmittableExtrinsic<RuntimeConfig, OnlineClient>> {
+        const FORCE_BATCH: &str = "force_batch";
+
+        let call = dynamic::tx(UTILITY, FORCE_BATCH, vec![Value::from(transfers)]);
+        let (number, hash) = self
+            .finalized_head_number_and_hash()
+            .await
+            .context("failed to get the chain head while constructing a transaction")?;
+        let extensions = DefaultExtrinsicParamsBuilder::new()
+            .mortal_unchecked(number.into(), hash, block_hash_count.into())
+            .tip_of(0, self.database.properties().await.usd_asset.id());
+
+        self.api
+            .tx
+            .create_signed_with_nonce(&call, signer, nonce, extensions.build())
+            .context("failed to create a transfer transaction")
+    }
+
+    async fn current_nonce(&self, account: &Account) -> Result<Nonce> {
+        self.api
+            .blocks
+            .at(fetch_best_block(&self.methods).await?)
+            .await
+            .context("failed to obtain the best block for fetching an account nonce")?
+            .account_nonce(account)
+            .await
+            .context("failed to fetch an account nonce by the API client")
+    }
+
+    async fn process_unpaid(
+        &self,
+        block: &Block<RuntimeConfig, OnlineClient>,
+        mut changes: InvoiceChanges,
+        hash: Hash,
+        invoice: Account,
+        price: Balance,
+    ) -> Result<()> {
+        let balance = self.balance(hash, &invoice).await?;
+
+        if let Some(_remaining) = balance.checked_sub(price) {
+            changes.invoice.status = InvoiceStatus::Paid(price);
+
+            let block_nonce = block
+                .account_nonce(&invoice)
+                .await
+                .context(BLOCK_NONCE_ERROR)?;
+            let current_nonce = self.current_nonce(&invoice).await?;
+
+            if current_nonce <= block_nonce {
+                let properties = self.database.properties().await;
+                let block_hash_count = properties.block_hash_count;
+                let signer = changes.invoice.signer(self.database.pair())?;
+
+                let transfers = vec![construct_transfer(
+                    &changes.invoice.recipient,
+                    price - EXPECTED_USDX_FEE,
+                    self.database.properties().await.usd_asset,
+                )];
+                let tx = self
+                    .batch_transfer(current_nonce, block_hash_count, &signer, transfers.clone())
+                    .await?;
+
+                // if let Some(a) = (fee + properties.existential_deposit + price).checked_sub(balance)
```

Comment:
maybe this stuff should be removed, not commented out - we do have version control here